File: src/Test_secant_root_solve.cpp
#include <catch.hpp>

#include <secant_root_solve.hpp>
#include <cmath>

double test_function(double x) {
    using namespace std;
    return sin(x);
}

TEST_CASE("Test secant_root_solve on a function", "[secant-root-solve]") {
    double a, b, truth, x;
    a = -1;
    b = 3;
    std::function<double(double)> func = test_function;
    truth = 0.0;
    x = numerics::secant_root_solve(func, a, b);
    REQUIRE( std::abs(x - truth) < 1e-12 );
}

File: Matlab/Test_secant_root_solve.m
%% Test_secant_root_solve.m
% 
% Test case for the Secant Root Solver function.
% 
% Use Graphical technique, bisection method, false-position, fixed-point
% iteration, Netwon method, and secant method to find the first root of
%     f(x) = x*exp(x) - cos(x)
% 
% @author: Matt Marti
% @date: 2019-04-26

clear

% Define function
f = @(x) cos(x) - x.*exp(x);


%% Part C: Secant Method

% Parameters
a = 0; % Lower bound
b = 1; % Upper bound
errstop = 1e-12; % Stopping criteria
maxiter = 1000;

% Function call
x = secant_root_solve(f, a, b, maxiter, errstop);

% Check results
assert(abs(f(x)) < errstop, 'Results error not less than specified error');

